landoapi/uplift.py

```diff
@@ -213,10 +221,76 @@ def create_approval_request(phab: PhabricatorClient, revision: dict, form_conten
     }
 
 
-def stack_uplift_form_submitted(stack_data) -> bool:
+def stack_uplift_form_submitted(stack_data: RevisionData) -> bool:
     """Return `True` if the stack has a valid uplift request form submitted."""
     # NOTE: this just checks that any of the revisions in the stack have the uplift form
     # submitted.
     return any(
         get_uplift_request_form(revision) for revision in stack_data.revisions.values()
     )
+
+
+def create_uplift_bug_update_payload(
+    bug: dict, repo_name: str, milestone: int
+) -> dict[str, Any]:
+    """Create a payload for updating a bug using the BMO REST API."""
+    payload: dict[str, Any] = {
+        "ids": [bug["id"]],
+    }
+
+    milestone_tracking_flag = f"cf_status_firefox{milestone}"
+    if "leave-open" not in bug["keywords"] and milestone_tracking_flag in bug:
+        # Set the status of a bug to fixed if the fix was uplifted to a branch
+        # and the "leave-open" keyword is not set.
+        payload[milestone_tracking_flag] = "fixed"
+
+    checkin_needed_flag = f"[checkin-needed-{repo_name}]"
+    if checkin_needed_flag in bug["whiteboard"]:
+        # Remove "[checkin-needed-beta]" etc. texts from the whiteboard.
+        payload["whiteboard"] = bug["whiteboard"].replace(checkin_needed_flag, "")
+
+    return payload
+
+
+UPDATE_RETRIES = 3
+
+
+def update_bugs_for_uplift(
+    bmo_url: str,
+    bmo_api_key: str,
+    changeset_titles: list[str],
+    repo_name: str,
+    milestone: int,
+):
+    """Update Bugzilla bugs for uplift."""
+    headers = {"X-Bugzilla-API-Key": bmo_api_key}
+    bugs = [str(bug) for title in changeset_titles for bug in parse_bugs(title)]
+    params = {
+        "ids": ",".join(bugs),
+    }
+    bug_endpoint = f"{bmo_url}/rest/bug"
+    resp_get = requests.get(bug_endpoint, headers=headers, params=params)
+    resp_get.raise_for_status()
+
+    bugs = resp_get.json()["bugs"]
+
+    for bug in bugs:
+        payload = create_uplift_bug_update_payload(bug, repo_name, milestone)
+
+        for i in range(1, UPDATE_RETRIES + 1):
+            # Update bug and account for potential errors.
+            try:
+                resp = requests.put(bug_endpoint, headers=headers, json=payload)
+                resp.raise_for_status()
+
+                continue
+            except requests.RequestException as e:
+                if i == UPDATE_RETRIES:
+                    raise e
+
+                logger.exception(
+                    f"Error while updating bugs after uplift on attempt {i}, retrying..."
+                )
+                logger.exception(str(e))
+
+                time.sleep(0.1 * i)
```

Comment:
Is this meant to be updated before merging this PR? These requests will still happen very quickly, even on the last attempt.